logstash-core/lib/logstash/pipeline.rb

```diff
@@ -12,20 +12,32 @@
 require "logstash/config/cpu_core_strategy"
 require "logstash/util/defaults_printer"
 require "logstash/shutdown_controller"
+require "logstash/util/wrapped_synchronous_queue"
+require "logstash/pipeline_reporter"
 
 module LogStash; class Pipeline
-  attr_reader :inputs, :filters, :outputs, :input_to_filter, :filter_to_output
+  attr_reader :inputs, :filters, :outputs, :worker_threads, :events_consumed, :events_filtered, :reporter, :pipeline_id
+
+  DEFAULT_SETTINGS = {
+    # We don't set pipeline workers defaults here, but rather in safe_pipeline_workers
```

Comment:
This is VERY confusing. I agree. The comment should be changed.

That being said, the whole idea of safe pipeline workers is temporary and is only here to not change too much compared to the current pipeline. I plan on submitting a patch that removes the whole idea of 'safe' pipeline workers by just using a mutex around filters that aren't thread safe. I've already opened a PR about it #4272.  That should simplify the whole situation.